lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go

```diff
@@ -89,11 +95,96 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		subProto.NetworkId = &protos.NetworkID{Id: networkID}
 		subProtos = append(subProtos, subProto)
 	}
+
 	listRes := &lte_protos.ListSubscribersResponse{
 		Subscribers:   subProtos,
 		NextPageToken: nextToken,
 	}
-	return listRes, nil
+
+	return listResWithDigest(s.flatDigestEnabled, listRes, req, gateway, apnsByName, apnResourcesByAPN), nil
+}
+
+// listResWithDigest returns a ListSubscribersResponse with the fields augmented
+// according to the flat digest logic.
+func listResWithDigest(
+	flatDigestEnabled bool,
+	listRes *lte_protos.ListSubscribersResponse,
+	req *lte_protos.ListSubscribersRequest,
+	gateway *protos.Identity_Gateway,
+	apnsByName map[string]*lte_models.ApnConfiguration,
+	apnResourcesByAPN lte_models.ApnResources,
+) *lte_protos.ListSubscribersResponse {
+
+	listRes = &lte_protos.ListSubscribersResponse{
+		Subscribers:   listRes.Subscribers,
+		NextPageToken: listRes.NextPageToken,
+		Digest:        req.Digest,
+		NoUpdates:     false,
+	}
+
+	// This functionality is currently placed behind a feature flag.
+	if flatDigestEnabled {
+		if req.PageToken == "" {
+			digest, err := getDigest(gateway, apnsByName, apnResourcesByAPN)
+			// If digest generation fails, the error is swallowed to not affect the main functionality.
+			if err != nil {
+				glog.Errorf("Generating digest for subscribers in network of gateway %s failed", gateway.GetLogicalId())
+			} else {
+				listRes.Digest = &lte_protos.SubscriberDigest{Md5Base64Digest: digest}
+				listRes.NoUpdates = req.Digest.Md5Base64Digest == digest
+				if listRes.NoUpdates {
+					listRes.Subscribers = []*lte_protos.SubscriberData{}
+				}
+			}
+		}
+	}
+
+	return listRes
+}
+
+// getDigest loads all subscribers registered on the current network, and returns
+// a deterministic subscriber flat digest.
+func getDigest(
+	gateway *protos.Identity_Gateway,
+	apnsByName map[string]*lte_models.ApnConfiguration,
+	apnResourcesByAPN lte_models.ApnResources,
+) (string, error) {
+	networkID := gateway.GetNetworkId()
+	subProtosById := map[string]proto.Message{}
+	curPageToken := ""
+
+	for {
+		lc := configurator.EntityLoadCriteria{
+			PageSize:           0,
+			PageToken:          curPageToken,
+			LoadConfig:         true,
+			LoadAssocsToThis:   true,
+			LoadAssocsFromThis: true,
+		}
+		subEnts, nextPagetoken, err := configurator.LoadAllEntitiesOfType(
+			networkID, lte.SubscriberEntityType, lc, serdes.Entity,
+		)
+		if err != nil {
+			return "", errors.Wrapf(err, "load all subscribers in network of gateway %s", networkID)
+		}
+		for _, sub := range subEnts {
+			subProto, err := convertSubEntsToProtos(sub, apnsByName, apnResourcesByAPN)
+			if err != nil {
+				return "", err
+			}
+			subProto.NetworkId = &protos.NetworkID{Id: networkID}
+
+			index := subProto.Sid.Id
+			subProtosById[index] = subProto
+		}
+
+		if nextPagetoken == "" {
+			break
+		}
+		curPageToken = nextPagetoken
+	}
```

Comment:
This logic is redundant to the per-page logic, coalesce it into a fn (which we'll then also use in the subscriberdb_cache service)